### PR TITLE
[docs] Add a doc for scraping receivers

### PR DESCRIPTION
**Pull request by dmitryax**
This document adds some definitions, stability requirements and guidelines on how to update emitted metrics for scraping receivers.

### CODE REVIEW

**Comment by dmitryax** [comment]
@open-telemetry/collector-contrib-approvers PTAL

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:17]:
```suggestion
some cases it may be impossible to fetch particular metrics from a system in a particular state. If it's known that 
```

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:19]: Is there a format that's recommended for this? It would be great to be consistent

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:21]: Are there additional categories? If not, i would remove the work `main` here

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:23]: Any thoughts around naming these `Enabled` and `Disabled` metrics? The behaviour between these two is opposite, but the word default doesn't imply it's the opposite of optional

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:30]:
```suggestion
treated as metrics essential to determine healthiness of a particular monitored entity. Optional metrics on the 
```

File: docs/scraping-receivers.md
Comment (codeboten) [inline docs/scraping-receivers.md:60]: Please link the "changing emitted metrics" section below here

**Review by codeboten — COMMENTED** [review]
Thank you for documenting this, just a few comments/questions

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:23]: I'm not sure. For me, it feels like Default/Optional are more intuitive even if Default doesn't imply it's the opposite of Optional. But I'm good with Enabled/Disabled as well. The only concern is that Enabled is the same name the settings option which may be ok. 

Let's keep it for now, I can rename them if there is no other feedback from @open-telemetry/collector-contrib-approvers

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: removed

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: done

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:19]: I don't know what format we can prescribe here. Conditions can vary significantly. Do you want to add an example like "The metric is only emitted for Redis version >= 6"?

File: docs/scraping-receivers.md
Comment (djaglowski) [reply docs/scraping-receivers.md:23]: Another suggestion: `DefaultOn` / `DefaultOff`. It's more verbose but it makes clear that we are describing the default state only.

`Enabled` is problematic because then when the user disables the metric you have a "disabled Enabled metric"?

`Optional` is also problematic because all metrics are optional.

File: docs/scraping-receivers.md
Comment (djaglowski) [inline docs/scraping-receivers.md:62]: Whether or not these metric names should be tracked is not settled in the specification itself.

https://github.com/open-telemetry/opentelemetry-specification/issues/2610

File: docs/scraping-receivers.md
Comment (djaglowski) [inline docs/scraping-receivers.md:77]: Some backends would consider this breaking because metrics would become associated with a new resource.

File: docs/scraping-receivers.md
Comment (djaglowski) [inline docs/scraping-receivers.md:123]: If a metric is being replaced with another, I think we should just refer to "Removing a default metric" and "Adding a default metric"?

That said, I think we should consider whether there is a reasonable process for making breaking changes to a metric (without changing its name). Often, a metric name is very intuitive and really seems like the "right" name for that metric. I think we're creating an undesirable situation if metric names are burned for good due to bugs in the data model that are resolved during beta.

File: docs/scraping-receivers.md
Comment (djaglowski) [inline docs/scraping-receivers.md:137]: I'm not sure this is really necessary or even reasonable.

With any other signal type, you would expect to get extra data that you do not care about, and either filter it or just ignore it.

Over time metrics will be added for many reasons and our scrapers will become less and less user friendly. 

To separate the concerns listed here:
- Are there backends that are actually "broken" by unexpected metrics?
- As long as additional load is roughly linear, is this really a problem?

**Review by djaglowski — COMMENTED** [review]
Thanks for writing this up. We definitely need to pin these details down.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:3]: Makes sense. I'll add the some categorization of scraping receivers. "Predefined scraping metrics receiver" sounds a bit vague. Maybe "Built-in scraping metrics receiver"?

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:31]: Thanks @djaglowski for providing the list. I agree that we should define it as a minimal requirement for marking a metric as default. But I don't agree that we shouldn't even try to consider "usefulness" of a metric before picking up the category. I'm afraid that if we go with this strict set of rules, a user that needs a rare metric which isn't not needed to anyone else will be encouraged to add it as default one, that will make other users to disable more and more metrics in their settings. It can be especially disruptive over time when receiver is marked beta or stable.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:62]: Ok I'll remove this requirement for now

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: Makes sense, replaced with zookeeper receiver

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:86]: I'll remove this line for now. But potentially we can consider an ability to hide non-identifying resource attribute in user settings. In that case it'll become:
```
- Adding a new optional resource attribute.
```

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:146]: > Are there backends that are actually "broken" by unexpected metrics?
> As long as additional load is roughly linear, is this really a problem?

Not "broken" but can add an unexpected increase in billing at least. For example, in Splunk Infrastructure Monitoring, some metrics are categorized by use case and billed by number of identified entities, e.g. number of hosts or containers. Each category has a predefined set of metrics. Other metrics are going to a "custom" category which is billed by a number of time series. If we don't consider it as a breaking change and don't notify users about a new metric that will be enabled by default, they will get an unexpected increase in the custom time series affecting their billing.

Comment (dmitryax) [reply docs/scraping-receivers.md:146]: > Over time metrics will be added for many reasons and our scrapers will become less and less user friendly.

Not sure I understand why scrapers will become less and less user friendly?

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: Makes sense, updated

**Comment by dmitryax** [comment]
> Tangential to this PR, we should also update https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/mdatagen/README.md to be more accurate. Maybe linking to this doc would be enough?

I believe https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/mdatagen/README.md should describe technical aspects of using the mdatagel tool. I'll update it once this PR is merged

File: docs/scraping-receivers.md
Comment (djaglowski) [reply docs/scraping-receivers.md]: That would be nice. In general the lack of distinction between identifying and non-identifying resource attributes is problematic.

File: docs/scraping-receivers.md
Comment (djaglowski) [reply docs/scraping-receivers.md:145]: > Not sure I understand why scrapers will become less and less user friendly?

Let's say we enable N metrics for a scraper at the time it is marked stable, over time this set of metrics will become outdated, often through events that are outside of our control:
- Source technologies will deprecate and remove metrics from their APIs
- Source technologies will add new metrics, some of which are critical
- We will realize that we missed important metrics
- We will occasionally make a mistake and realize that a metric we added is problematic (e.g. causes excessive load on the source)

Over time, the set of metrics that we launched with grows stale, and in order for users to find the scraper useful, they will have to make more and more modifications to the default set of metrics. 

I think it would be unfortunately if we are unable to make sensible adjustments to the scraper's output from time to time. In my opinion, metrics should each have their own lifecycle so that our scrapers are not frozen in the past forever. This model would reflect the reality of the situation much better. We should look for another way to define and control a static _set_ of metrics.

Comment (dmitryax) [reply docs/scraping-receivers.md:145]: I completely agree that the set of default metrics should be always up to date and we shouldn't freeze them when we mark the component stable. That's why I mentioned that adding/removing metrics is possible in the stable state, but I don't think we should promote doing it silently. What if we relieve the requirement for such changes in the beta/stable states and mention all the acceptable reasons that you provided? Or you don't agree that we should call them "breaking" at all?

Comment (djaglowski) [reply docs/scraping-receivers.md:145]: > That's why I mentioned that adding/removing metrics is possible in the stable state, but I don't think we should promote doing it silently. What if we relieve the requirement for such changes in the beta/stable states and mention all the acceptable reasons that you provided?

Ok, I agree with what you're suggesting here. We can keep the metrics up to date by _expecting_ breaking changes, while also communicating them in a way that is easily noticed.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:31]: Great idea! Submitted an issue for this improvement for later https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15995

Comment (dmitryax) [reply docs/scraping-receivers.md:31]: I'll remove this requirement from the doc for now

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:124]: Updated the definition of stable to expect breaking changes in order to keep the metrics set up to date.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: Makes sense. Thanks for the feedback. I removed the "replacement" sections to not encourage unnecessary renames. Also added a section for other in-place changes using feature gates.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: @djaglowski I incorporated the guidelines that you suggested in the doc. PTAL

**Comment by dmitryax** [comment]
@djaglowski @codeboten @evan-bradley thanks for your feedback. I believe everything is addressed now. Please take another look

**Review by djaglowski — APPROVED** [review]
LGTM. Thanks for considering my feedback.

File: docs/scraping-receivers.md
Comment (astencel-sumo) [inline docs/scraping-receivers.md:13]:
```suggestion
- **Built-in scraping metrics receivers:** Receivers of this type emit a predefined set of metrics. However, the 
```

**Review by astencel-sumo — APPROVED** [review]
This is very educative read, thanks a lot Dmitrii.

My recent [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14927) is a good example of applying the "Changing the emitted metrics" section. One thought related to this is that I will need to add the warning messages after this doc is accepted.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md]: Good catch. Thanks!

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:55]:
```suggestion
All the requirements defined for components in [the Collector's README](../README.md#stability-levels) are applicable to 
```

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:91]:
```suggestion
than 0.65.0. Any warnings SHOULD include the version starting from which the next step will take effect. If a
```

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:123]:
```suggestion
Adding a new default metric is a breaking change for a scraping receiver because it introduces an unexpected output for users 
```

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:124]:
```suggestion
and additional load on metric backends. Steps to apply such a change:
```

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:135]:
```suggestion
example, if a metric has to be renamed for any reason, the guidelines for "Removing an optional metric" and "Adding a 
```

File: docs/scraping-receivers.md
Comment (evan-bradley) [inline docs/scraping-receivers.md:134]:
```suggestion
Other breaking changes SHOULD follow similar strategies that modify the `enabled` field in user settings. For 
```

**Review by evan-bradley — APPROVED** [review]
Thanks for writing this up.

I left a few wordsmithing suggestions, but overall it looks good to me.

File: docs/scraping-receivers.md
Comment (dmitryax) [reply docs/scraping-receivers.md:134]: This correction changes the statement. We don't modify `enabled` field in the settings, we check if it's set only

**Comment by dmitryax** [comment]
@evan-bradley thanks for the review! I applied all the corrections except for the last one

**Review by bogdandrutu — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
